core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -381,9 +404,19 @@ private void cleanUncommittedAppends(Set<ManifestFile> committed) {
       this.cachedNewManifest = null;
     }
 
-    for (ManifestFile manifest : appendManifests) {
-      if (!committed.contains(manifest)) {
-        deleteFile(manifest.path());
+    // the clean-up logic depends on whether we append original manifests or their copies
+    // 1) if we rewrite manifests before appending them
+    //    we need to delete all uncommitted copies of manifests to prevent orphan metadata files
+    //    as the caller doesn't have access to them
+    // 2) if we append original manifests without rewriting them and the commit succeeds
+    //    we need to delete appended manifests that are merged with others while preparing a new snapshot
+    //    otherwise, the caller will have to find which appended manifests were merged but not committed
+    //    to prevent orphan metadata files
+    if (!snapshotIdInheritanceEnabled || !committed.isEmpty()) {
```

Comment:
`appendManifests` always contains manifests that are owned by the table. If inheritance is not enabled, then the manifest is rewritten before it is added to appendManifests, and if it is enabled then the table takes ownership of the manifest.

So it should always be okay to clean up `appendManifests` using the committed list here, regardless of the inheritance setting.

Reply:
I was wrong: there's one case where we need to avoid cleanup: when the commit failed and inheritance is enabled.

I think that explains the check here. If inheritance is not enabled or if the committed list is empty, then we know that is not the case. The logic here is difficult to follow, even with the comments, but I think it is correct.

Reply:
I think it would be easier to think about this if we kept owned manifests and non-owned manifests separate. Then we would be able to have the following logic here:

```java
  // appendManifests are rewritten and are owned by the table
  for (ManifestFile manifest : appendManifests) {
    if (!committed.contains(manifest)) {
      deleteFile(manifest.path());
    }
  }

  // inheritingManifests are only owned by the table if the commit succeeded
  if (!committed.isEmpty()) {
    // the commit succeeded if at least one manifest was committed
    // the table now owns the inheritingManifests; clean up any that are not used
    for (ManifestFile manifest : inheritingManifests) {
      if (!committed.contains(manifest)) {
        deleteFile(manifest.path());
      }
    }
  }
```

@aokolnychyi, what do you think?

Reply:
I've implemented an approach that is similar to the snippet above. I'll need to do one more pass to make sure I didn't forget anything.